localstack/services/s3/s3_listener.py

```diff
@@ -1307,5 +1320,91 @@ def return_response(self, method, path, data, headers, response, request_handler
+    AWS_ACCESS_KEY = 'temp'
+    AWS_ACCESS_SECRET_KEY = 'temp'
+
+    # Checking required parameters are present in url or not
+    if 'Signature' not in query_params or 'Expires' not in query_params or 'AWSAccessKeyId' not in query_params:
+        return requests_error_response_xml_presign_url_auth(
+            code=403,
+            message='Query-string authentication requires the Signature, Expires and AWSAccessKeyId parameters',
+            code_string='AccessDenied'
+        )
+
+    # Fetching headers which has been sent to the requets
+    for header in headers:
+        key = header[0]
+        if key not in ignored_headers:
+            sign_headers.append(header)
+
+    # Request's headers are more essentials than the query parameters in the requets.
+    # Different values of header in the header of the request and in the query paramter of the requets url
+    # will fail the signature calulation. As per the AWS behaviour
+    presign_query_params = ['Signature', 'Expires', 'AWSAccessKeyId']
+    if len(query_params) > 2:
+        for key in query_params:
+            if key not in presign_query_params:
+                if key.lower() not in (header[0].lower() for header in headers):
+                    sign_headers.append((key, query_params[key][0]))
+
+    # Preparnig dictionary of request to build AWSRequest's object of the botocore
+    request_dict = {
+        'url_path': path.split('?')[0],
+        'query_string': {},
+        'method': method,
+        'headers': dict(sign_headers),
+        'body': b'',
+        'url': url.split('?')[0],
+        'context': {
+            'is_presign_request': True,
+            'use_global_endpoint': True,
+            'signing': {
+                'bucket': str(path.split('?')[0]).split('/')[1]
+            }
+        }
+    }
+    aws_request = create_request_object(request_dict)
+
+    # Calculating Signature
+    credentials = Credentials(access_key=AWS_ACCESS_KEY, secret_key=AWS_ACCESS_SECRET_KEY)
+    auth = HmacV1QueryAuth(credentials=credentials, expires=query_params['Expires'][0])
+    split = urlsplit(aws_request.url)
+    string_to_sign = auth.get_string_to_sign(method=method, split=split, headers=aws_request.headers)
+    signature = auth.get_signature(string_to_sign=string_to_sign)
+
+    # Comparing the signature in url with signature we calculated
+    if query_params['Signature'][0] != signature:
+        return requests_error_response_xml_presign_url_auth(
+            code=403,
+            code_string='SignatureDoesNotMatch',
+            aws_access_token=AWS_ACCESS_KEY,
+            string_to_sign=string_to_sign,
+            signature=signature,
+            message='The request signature we calculated does not match the signature you provided. \
+                    Check your key and signing method.')
+
+    # Checking whether the url is expired or not
+    if query_params['Signature'][0] == signature and int(query_params['Expires'][0]) < time.time():
```

Comment:
The first check (`query_params['Signature'][0] == signature`) should be obsolete (already checked in line 1389). Can be simplified to:
```
if int(query_params['Expires'][0]) < time.time():
```

Reply:
Sure